panther_lights/src/apa102.cpp

```diff
@@ -0,0 +1,124 @@
+// Copyright 2023 Husarion sp. z o.o.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <panther_lights/apa102.hpp>
+
+#include <fcntl.h>
+#include <linux/spi/spidev.h>
+#include <sys/ioctl.h>
+#include <unistd.h>
+
+#include <cmath>
+#include <cstdint>
+#include <cstring>
+#include <fstream>
+#include <stdexcept>
+#include <string>
+#include <vector>
+
+namespace panther_lights::apa102
+{
+
+APA102::APA102(const std::string & device, const std::uint32_t speed, const bool cs_high)
+: fd_(open(device.c_str(), O_WRONLY)), device_(device), speed_(speed)
+{
+  if (fd_ < 0) {
+    throw std::ios_base::failure("Failed to open " + device_);
+  }
+
+  static std::uint8_t mode = cs_high ? SPI_MODE_3 : SPI_MODE_3 | SPI_CS_HIGH;
+  if (ioctl(fd_, SPI_IOC_WR_MODE32, &mode) == -1) {
+    close(fd_);
+    throw std::ios_base::failure("Failed to set mode for " + device_);
+  }
+
+  if (ioctl(fd_, SPI_IOC_WR_BITS_PER_WORD, &kBits) == -1) {
+    close(fd_);
+    throw std::ios_base::failure("Can't set bits per word for " + device_);
+  }
+
+  if (ioctl(fd_, SPI_IOC_WR_MAX_SPEED_HZ, &speed_) == -1) {
+    close(fd_);
+    throw std::ios_base::failure("Can't set speed for " + device_);
+  }
+}
+
+APA102::~APA102() { close(fd_); }
+
+void APA102::SetGlobalBrightness(const double brightness)
```

Comment:
```suggestion
void APA102::SetGlobalBrightness(const float brightness)
```
`panther_msgs/srvs/SetLEDBrightness.data` if float so I don't see a point of using double in here